Redirecting to a variable should be possible

JamesWTruher opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

There are a number of situations where it would be very convenient to _redirect_ data to a variable using the `variable:name` syntax. This PR provides this behavior by inspecting the target of the redirection and if it uses the variable provider will call `Set-Variable` rather than `Out-File`.

With this PR the following scenarios are possible:

```powershell
PS /> .{                                                                                   
>> "Output 1"
>> Write-Warning "Warning, Warning!"
>> "Output 2"
>> } 3> variable:warns
Output 1
Output 2
PS /> $warns
WARNING: Warning, Warning!
PS /> 
```

This could prove very helpful to some logging scenarios as well.

```powershell
S /> .{                                                                                   
>> "Output 1"
>> Write-Warning "This is a warning"
>> "Output 2"
>> Write-Debug -Debug "and a debug message"
>> "Output 3"
>> } *> variable:kitchenSync
PS /> $kitchenSync
Output 1
WARNING: This is a warning
Output 2
DEBUG: and a debug message
Output 3
```

Append is also supported via `>>`

```powershell
PS /> .{ "output 1" } *>variable:outputs                                                   
PS /> .{ "output 2" } *>>variable:outputs
PS /> .{ write-warning warn! } *>>variable:outputs
PS /> .{ "output 3" } *>>variable:outputs         
PS /> $outputs                   
output 1
output 2
WARNING: warn!
output 3
```

and redirecting multiple streams to multiple targets, based on the stream number:

```powershell
PS /> .{
>> "output 1"
>> write-warning "Warning!"
>> "output 2"
>> write-debug -debug "debug here"
>> "output 3"
>> write-information Info
>> "output 4"
>> write-error "That's bad!"
>> Write-Verbose -Verbose "and verbose"
>> "all done"
>> } > variable:outputStream 2>variable:errorStream 3>variable:warningStream 4>variable:verboseStream 5>variable:debugStream 6>variable:InfoStream
PS /> $outputStream
output 1
output 2
output 3
output 4
all done
PS /> $errorStream 
Write-Error: That's bad!
PS /> $warningStream
WARNING: Warning!
PS /> $verboseStream                   
VERBOSE: and verbose
PS /> $debugStream  
DEBUG: debug here
PS /> $InfoStream 
Info
```

Lastly, object-ness is preserved:

```powershell
PS /> .{                                 
>> "output"
>> write-error "bad"
>> write-warning "warn"
>> write-verbose -verbose "verbose"
>> write-debug -debug debug
>> } *>variable:all
PS /> $all | %{$_.gettype()}

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     String                                   System.Object
True     False    ErrorRecord                              System.Object
True     False    WarningRecord                            System.Management.Automation.InformationalRecord
True     False    VerboseRecord                            System.Management.Automation.InformationalRecord
True     False    DebugRecord                              System.Management.Automation.InformationalRecord
```


This is not to say that redirection to a variable should take the place of assignment, but it does enable some new scenarios.

the current behavior when redirecting to a variable is an error (because redirection is implemented via `Out-File`):

```powershell
PS /> "sldfkj" >variable:no          
Out-File: Cannot open file because the current provider (Microsoft.PowerShell.Core\Variable) cannot open a file.
```

<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name(s): `PSRedirectToVariable`
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

JamesWTruher committed:
add tests

JamesWTruher committed:
fix formatting issues

JamesWTruher committed:
Make redirection to variable an experimental feature.

Update test to check for experimental feature status.
Set-Variable will still have the -append parameter, but if used when the experimental feature is disabled, a parameter binding error will result.

JamesWTruher committed:
Check to be sure provider is not null.

JamesWTruher committed:
update to use different FullyQualifiedErrorId

JamesWTruher committed:
use Experimental attribute for append parameter rather than runtime check.

JamesWTruher committed:
Revert "update to use different FullyQualifiedErrorId"

This reverts commit 8b34af1ac19ecdb821b34534e446ab129830c92f.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs:
```diff
@@ -710,6 +717,32 @@ protected override void BeginProcessing()
+                {
+                    ErrorRecord appendVariableError = new ErrorRecord(new InvalidOperationException(), "SetVariableAppend", ErrorCategory.InvalidOperation, Name);
+                    appendVariableError.ErrorDetails = new ErrorDetails("SetVariable");
+                    appendVariableError.ErrorDetails.RecommendedAction = "Use a single variable rather than a collection";
```
Should it be in Resx file?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs:
```diff
@@ -710,6 +717,32 @@ protected override void BeginProcessing()
+
+                _valueList = new List<object>();
+                var currentValue = Context.SessionState.PSVariable.Get(Name[0]);
+                if (currentValue != null) {
+                    if (currentValue.Value is IList<object>)
+                    {
+                        _valueList.AddRange(currentValue.Value as IList<object>);
+                    }
```
```suggestion
                if (currentValue is not null) {
                    if (currentValue.Value is IList<object> ilist)
                    {
                        _valueList.AddRange(ilist);
                    }
```

JamesWTruher committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs

Co-authored-by: Ilya <darpa@yandex.ru>

JamesWTruher committed:
Move remediation steps into resource.

Update Set-Variable -Append to be correct when -name and -value are used.
Add tests for the new behavior in Set-Variable.

JamesWTruher committed:
Change expected error for contrained language mode and redirection.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs:
```diff
@@ -692,6 +692,13 @@ public SwitchParameter PassThru
         private bool _passThru;
 
+        /// <summary>
+        /// Append to the variable if it exists
```
```suggestion
        /// Append to the variable if it exists.
```

JamesWTruher committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs

Co-authored-by: Ilya <darpa@yandex.ru>

JamesWTruher committed:
Support redirection of native app.

Add tests to validate.

SeeminglyScience reviewed (approved):
Some very minor style nits, other than that looks great! Very excited for this change ❤️

JamesWTruher committed:
testexe needs proper case to run on linux.

JamesWTruher committed:
Address codefactor issues 01.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs:
```diff
@@ -693,6 +693,13 @@ public SwitchParameter PassThru
+        /// Gets whether we will append to the variable if it exists.
+        /// </summary>
+        [Parameter]
+        [Experimental("PSRedirectToVariable", ExperimentAction.Show)]
```
```suggestion
        [Experimental(ExperimentalFeature.PSRedirectToVariable, ExperimentAction.Show)]
```

You already have the const string defined in this class

SteveL-MSFT commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -1090,10 +1090,23 @@ public override string ToString()
         {
+            // Check first to see if File is a variable path. If so, we'll not create the FileBytePipe
+            bool redirectToVariable = false;
+            if (ExperimentalFeature.IsEnabled("PSRedirectToVariable"))
```
```suggestion
            if (ExperimentalFeature.IsEnabled(ExperimentalFeature.PSRedirectToVariable))
```

SteveL-MSFT commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -1208,26 +1221,51 @@ internal Pipe GetRedirectionPipe(ExecutionContext context, PipelineProcessor par
-                /*argumentAst*/null, File,
-                false);
-            commandProcessor.AddParameter(cpi);
+            if (ExperimentalFeature.IsEnabled("PSRedirectToVariable") && p != null && p.NameEquals(context.ProviderNames.Variable))
```
```suggestion
            if (ExperimentalFeature.IsEnabled(ExperimentalFeature.PSRedirectToVariable) && p != null && p.NameEquals(context.ProviderNames.Variable))
```

SteveL-MSFT commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -1208,26 +1221,51 @@ internal Pipe GetRedirectionPipe(ExecutionContext context, PipelineProcessor par
-            // Now, We can take what ever has been set if PSDefaultParameterValues
-            // Unicode is still the default, but now may be overridden
+            // determine whether we're trying to set a variable by inspecting the file path
+            // if we can determine that it's a variable, we'll use Set-Variable rather than Out-File
```
Just a note and not something you need to change, but it seems in the future, it would make more sense to have `Out-Provider` and have the provider subsystem determine which provider receives it rather than hardcoding to specific cases here

JamesWTruher replied:
I did think about that when I put this together and thought that the scenarios for redirection to something other than filesystem or variable were pretty limited, so I took this approach.

SteveL-MSFT replied:
I agree that a smaller change now makes sense.  Created https://github.com/PowerShell/PowerShell/issues/23794 to capture this.

JamesWTruher committed:
Update src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs

Co-authored-by: Steve Lee <slee@microsoft.com>

JamesWTruher committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/Var.cs

Co-authored-by: Steve Lee <slee@microsoft.com>

JamesWTruher committed:
Update src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs

Co-authored-by: Steve Lee <slee@microsoft.com>
